Allow MOOSE_DIR to be treated like an app

milljm opened the pull request:
When calling versioner.py, no longer short circuit the get_app method if ran while inside MOOSE_DIR. Instead, return with a special identifier designating we want to build the combined-opt application.

`./versioner.py moose` remains valid
`./versioner.py app` while inside MOOSE_DIR will no longer fail

Commit:
Allow MOOSE_DIR to be treated like an app

When calling versioner.py, no longer short circuit the get_app method
if ran while inside MOOSE_DIR. Instead, return with a special
identifier designating we want to build the combined-opt application.

`./versioner.py moose` remains valid
`./versioner.py app` while inside MOOSE_DIR will no longer fail

Closes #23247

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1317658/) on e8b1e1d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23250/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1317672/) on e8b1e1d wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23250/coverage/xfem)

</details>



This comment will be updated on new commits.

loganharbour commented:
@milljm glad I added https://civet.inl.gov/job/1317845/ 😆

milljm commented:
Now that I have a bunch of CIVET_HEAD_SHA, etc... I can test it :)
I gave this -9 chance of success...

The unittests though, I didn't expect!

Commit:
Add vscode to gitignore

Ignore .vscode data

Commit:
Update unittest, fix jinja app.def template

Now that we return proper variables, test MOOSE_DIR app method in
the unittest.

Correct comparison operators in jinja template.

loganharbour reviewed (changes requested):
While we're at it, can you update https://github.com/idaholab/moose/blob/next/scripts/tests/versioner_hashes.yaml from https://github.com/idaholab/moose/pull/23243?

Commit:
Adding hashes to versioner_hashes.yaml

Will add a Precheck to Civet recipes to reflect the comments being
added.

Commit:
Name the container moose-combined

Changes necessary to support naming scheme.

Commit:
Correct binary name in unittest

loganharbour commented:
@milljm what's the status here?

milljm commented:
This can merge... I can build the recipes that make use of it later/after.

milljm commented:
Investigating additional changes to allow template includes from other projects (Apps)

Commit:
Allow applications to modify app.def

Allow applications with an 'apptainer' directory to change designated
sections/areas in moose/apptainer/app.def when generating def files
using apptainer_generator.py.

Commit:
Clean the code a little bit

Commit:
Update latest hash for this PR

who would of thunk it would break unittests.

Commit:
Fix mistakes

Syntax and other mistake, fixes

milljm commented:
Adding fine-grained sectional replacement enhancements...

Commit:
finer grained sectional replacements

allow for finer replacment in app.def

Commit:
Adding test_pre_test (Cardinal cross sections)

The necessary bits to make Cardinal Tests pass

Commit:
Add allow_to_fail in app.def jinja template

Allow the building of a container to continue (within reason) if
there was a Civet recipe related failure, and that recipe is allowed
to fail.

Commit:
update version_hashes.yaml

loganharbour commented on scripts/versioner.py:
```diff
@@ -227,6 +223,10 @@ def get_app():
         app_name = os.path.basename(git_root).rstrip().lower()
 
+        # If we're nested within MOOSE_DIR, we're moose_combined app
+        if app_name == 'moose':
```
The more I think about this, this'll fail if someone calls their directory not "moose" or "moose.git". Do we care?

milljm replied:
we can strip off the .git if its there, but yeah I am not sure what to do if the user decides to rename the moose dir to something else entirely...

loganharbour commented on scripts/apptainer_generator.py:
```diff
@@ -346,6 +349,72 @@ def _add_definition_labels(self, definition):
         return definition
 
+    @staticmethod
+    def create_filename(app_root, section_key, actions):
```
Beautiful 🥳

loganharbour commented on apptainer/app.def:
```diff
@@ -110,33 +163,60 @@ From: {{ APPTAINER_FROM }}
         cd $TEMP_LOC
         ${BINARY_NAME}-${METHOD} --copy-inputs $TEST_DIR
         cd ${BINARY_NAME}/${TEST_DIR}
+{%- if allowed_to_fail is defined %}
+        # CIVET allowed_to_fail is True
+        set +e
+{%- endif %}
```
Also 🥳

loganharbour replied:
Oh - can you add an echo here that says you're allowing things to fail or not?

loganharbour reviewed (commented):
Looks good to me. Let me know when you want it in
